File: Server/Models/Result.cs
namespace WebServer.Models;

public sealed record Error(string Code, string? Description);

public class Result<T>
{
    private Result(bool isSuccess, Error? error = null, T? value = default)
    {
        IsSuccess = isSuccess;
        Error = error;
        Value = value;
    }

    public T? Value { get; }

    public bool IsSuccess { get; }

    public bool IsFailure => !IsSuccess;

    public Error? Error { get; }

    public static Result<T> Success(T value) => new(true, null, value);

    public static Result<T> Failure(Error error) => new(false, error, default);
}

public class Result
{
    public bool IsSuccess { get; }
    public bool IsFailure => !IsSuccess;
    public Error? Error { get; }

    protected Result(bool isSuccess, Error? error = null)
    {
        IsSuccess = isSuccess;
        Error = error;
    }

    public static Result Failure(Error error) => new(false, error);

    public static Result Success() => new(true);
}

File: Client/src/app/shared/Util/Result.ts
export type Failure = {
  _tag: 'Failure';
  success: false;
  error: Error;
  data?: never;
};

export type Success<T = void> = {
  _tag: 'Success';
  success: true;
  data: T;
};

export type Error = {
  code: string;
  message: string;
};

export type Result<T = void, E = unknown> = Failure | Success<T>;

export function succeed<T = void>(data: T): Success<T> {
  return {
    _tag: 'Success',
    data: data,
    success: true,
  };
}

export function fail(error: Error): Failure {
  return {
    _tag: 'Failure',
    error: error,
    success: false,
  };
}
